File: ruby/password-validator.rb
class PasswordValidator
  def initialize(password)
    @password = password
    @strength = 0
  end

  def validate_length
    if @password.length >= 5 && @password.length <= 9
      true
    elsif @password.length >= 10
      @strength += 2
      true
    else
      false
    end
  end

  def validate_special_chars(special_chars)
    if special_chars.chars.any? { |char| @password.include?(char) }
      @strength += 1
      true
    else
      false
    end
  end

  def validate_numbers(numbers_chars)
    if numbers_chars.chars.any? { |char| @password.include?(char) }
      @strength += 1
      true
    else
      false
    end
  end

  def validate_spaces
    !@password.include?(" ")
  end

  def display_password
    puts "Your entered password: #{@password}"
  end

  def summary(special_chars, numbers_chars)
    display_password
    if !validate_spaces
      puts "YOUR PASSWORD IS INVALID [CONTAINS SPACES]"
    else
      puts validate_numbers(numbers_chars) ? "Password contains numbers." : "Password does not contain any numbers."
      puts validate_special_chars(special_chars) ? "Password contains at least one special character." : "Password does not contain any special character."
      puts "Password has a normal or strong length." if validate_length
      puts "SUMMARY: YOUR PASSWORD STRENGTH IS #{@strength * 25} / 100%\n"
    end
  end
end

# Manually input password
print "Input your password: "
user_input_password = gets.chomp

# Define character sets
LOWERCASE_CHARS = "abcdefghijklmnopqrstuvwxyz"
UPPERCASE_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
NUMBERS = "0123456789"
SYMBOLS = "!@#$%^&*_=+-/"

# Validate the user-input password
validator = PasswordValidator.new(user_input_password)
validator.summary(SYMBOLS, NUMBERS)

File: php/password-validator.php
<?php

class PasswordValidator {
    private $password;
    private $strength;

    public function __construct($password) {
        $this->password = $password;
        $this->strength = 0;
    }

    public function validateLength() {
        if (strlen($this->password) >= 5 && strlen($this->password) <= 9) {
            return true;
        } elseif (strlen($this->password) >= 10) {
            $this->strength += 2;
            return true;
        } else {
            return false;
        }
    }

    public function validateSpecialChars($specialChars) {
        if (strpbrk($this->password, $specialChars) !== false) {
            $this->strength += 1;
            return true;
        } else {
            return false;
        }
    }

    public function validateNumbers($numbersChars) {
        if (strpbrk($this->password, $numbersChars) !== false) {
            $this->strength += 1;
            return true;
        } else {
            return false;
        }
    }

    public function validateSpaces() {
        if (strpos($this->password, ' ') === false) {
            return true;
        } else {
            return false;
        }
    }

    public function displayPassword() {
        echo "Your entered password: {$this->password}\n";
    }

    public function summary($specialChars, $numbersChars) {
        $this->displayPassword();
        if (!$this->validateSpaces()) {
            echo "YOUR PASSWORD IS INVALID [CONTAINS SPACES]\n";
        } else {
            echo $this->validateNumbers($numbersChars) ? "Password contains numbers.\n" : "Password does not contain any numbers.\n";
            echo $this->validateSpecialChars($specialChars) ? "Password contains at least one special character.\n" : "Password does not contain any special character.\n";
            if ($this->validateLength()) {
                echo "Password has a normal or strong length.\n";
            }
            echo "SUMMARY: YOUR PASSWORD STRENGTH IS " . ($this->strength * 25) . " / 100%\n";
        }
    }
}

// Manually input password
echo "Input your password: ";
$userInputPassword = trim(fgets(STDIN));

// Define character sets
$LOWERCASE_CHARS = "abcdefghijklmnopqrstuvwxyz";
$UPPERCASE_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
$NUMBERS = "0123456789";
$SYMBOLS = "!@#$%^&*_=+-/";

// Validate the user-input password
$validator = new PasswordValidator($userInputPassword);
$validator->summary($SYMBOLS, $NUMBERS);

?>
